Spotifyd prints error, disconnects, but does not exit

My spotifyd repeatedly prints an error message after a few minutes:
"[ERROR] Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" }"

After this message, either (a) spotifyd exists, or (b) playing stops, and the daemon becomes invisible to Spotify apps. Also, spotifyd cannot be killed with Ctrl+C any more. So it does not even react to signals any more. #127 could be related.

Could it help to just handle this connection reset error more gracefully?

I am also experiencing this issue on Odroid C2. I'm using systemd unit to manage spotifyd, and after line `[ERROR] Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" }` is printed in `systemctl status spotifyd`, it still reports spotifyd as running, but other Spotify clients won't connect to it and eventually it will be dropped.

I can provide any additional information that will help with resolving this bug.

I have the same issue. Has it been resolved?

Dell XPS 15 9550 
OS: Ubuntu 16.04
Steps to reproduce:
- download v0.2.5
- run: ./spotifyd  --no-daemon
- play music from phone
- wait 5-6 minutes
- "ConnectionReset"

`13:26:46 [ERROR] Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" }`

I have the same issue. do you know why is happening?

Raspberry py
OS: Raspian 8.0
Steps to reproduce:

download v0.2.11
run: spotifyd as systemd
Using alsa volume controller.

"ConnectionReset"
spotifyd[903]: 21:39:09 [ERROR] Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" }

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -34,7 +34,7 @@ tokio-io = "0.1"
 tokio-signal = "0.1"
 url = "1.7"
 xdg = "2.2"
-librespot = { version = "0.1", default-features = false, features = ["with-tremor"] }
+librespot = { version = "0.1.1", default-features = false, features = ["with-tremor"] }
 
 [target."cfg(target_os = \"macos\")".dependencies]
 whoami = "0.7.0"

--- a/src/config.rs
+++ b/src/config.rs
@@ -372,6 +372,11 @@ pub struct SharedConfigValues {
     /// The device type shown to clients
     #[structopt(long, possible_values = &DEVICETYPE_VALUES, value_name = "string")]
     device_type: Option<DeviceType>,
+
+    /// Autoplay on connect
+    #[structopt(long)]
+    #[serde(default, deserialize_with = "de_from_str")]
+    autoplay: bool,
 }
 
 #[derive(Debug, Default, Deserialize)]
@@ -577,6 +582,7 @@ pub(crate) struct SpotifydConfig {
     pub(crate) shell: String,
     pub(crate) zeroconf_port: Option<u16>,
     pub(crate) device_type: String,
+    pub(crate) autoplay: bool,
 }
 
 pub(crate) fn get_internal_config(config: CliConfig) -> SpotifydConfig {
@@ -615,6 +621,8 @@ pub(crate) fn get_internal_config(config: CliConfig) -> SpotifydConfig {
     let device_id = device_id(&device_name);
 
     let normalisation_pregain = config.shared_config.normalisation_pregain.unwrap_or(0.0f32);
+
+    let autoplay = config.shared_config.autoplay;
 
     let device_type = config
         .shared_config
@@ -689,6 +697,7 @@ pub(crate) fn get_internal_config(config: CliConfig) -> SpotifydConfig {
         shell,
         zeroconf_port: config.shared_config.zeroconf_port,
         device_type,
+        autoplay,
     }
 }
 

--- a/src/main_loop.rs
+++ b/src/main_loop.rs
@@ -95,6 +95,7 @@ pub(crate) struct MainLoopState {
     pub(crate) player_config: PlayerConfig,
     pub(crate) session_config: SessionConfig,
     pub(crate) handle: Handle,
+    pub(crate) autoplay: bool,
     pub(crate) linear_volume: bool,
     pub(crate) running_event_program: Option<Child>,
     pub(crate) shell: String,
@@ -165,6 +166,7 @@ impl Future for MainLoopState {
 
                 let (spirc, spirc_task) = Spirc::new(
                     ConnectConfig {
+                        autoplay: self.autoplay,
                         name: self.spotifyd_state.device_name.clone(),
                         device_type: self.device_type,
                         volume: mixer.volume(),

--- a/src/setup.rs
+++ b/src/setup.rs
@@ -73,6 +73,7 @@ pub(crate) fn initial_state(
     let player_config = config.player_config;
     let session_config = config.session_config;
     let backend = config.backend.clone();
+    let autoplay = config.autoplay;
     let device_id = session_config.device_id.clone();
 
     #[cfg(feature = "alsa_backend")]
@@ -92,6 +93,7 @@ pub(crate) fn initial_state(
     let discovery_stream = discovery(
         &handle,
         ConnectConfig {
+            autoplay,
             name: config.device_name.clone(),
             device_type,
             volume: mixer().volume(),
@@ -161,6 +163,7 @@ pub(crate) fn initial_state(
         running_event_program: None,
         shell: config.shell,
         device_type,
+        autoplay,
     }
 }
 

